cpp/src/toppra/constraint.hpp

```diff
@@ -1,6 +1,213 @@
+     * be of size 1.
+     * */
+    bool constantF () const
+    {
+      return constantF_;
+    }
+
+    /// Dimension of \f$g\f$.
+    Eigen::Index nbConstraints () const
+    {
+      return k_;
+    }
+
+    /// Dimension of \f$a, b, c, v\f$.
+    Eigen::Index nbVariables () const
+    {
+      return m_;
+    }
+
+    /** Compute numerical coefficients of the given constraint.
+     *
+     *  \param[in] path The geometric path.
+     *  \param[in] gridpoints Vector of size N+1. Gridpoint use for discretizing path.
+     *
+     *  \param[out] a N+1 Vector of size m.
+     *  \param[out] b N+1 Vector of size m.
+     *  \param[out] c N+1 Vector of size m.
+     *  \param[out] F N+1 Matrix of shape (k, m). If LinearConstraint::constantF
+     *              is \c true, there is only one such Matrix.
+     *  \param[out] g N+1 Vector of size m.
+     *  \param[out] ubound Shape (N + 1, 2). See notes.
+     *  \param[out] xbound Shape (N + 1, 2). See notes.
+     *
+     * \note the output must be allocated to correct sizes prior to calling this
+     * function.
+     *
+     * \todo check constness
+     *
+     * */
+    void computeParams(const GeometricPath& path, const Vector& gridpoints,
+        Vectors& a, Vectors& b, Vectors& c,
+        Matrices& F, Vectors& g);
+
+    virtual std::ostream& print(std::ostream& os) const;
+
+  protected:
+    /**
+     * \param k number of inequality constraints.
+     * \param m number of internal variable (i.e. dimention of \f$v\f$).
+     * */
+    LinearConstraint(Eigen::Index k, Eigen::Index m)
+      : discretizationType_ (Collocation)
+      , k_ (k), m_ (m)
+      , constantF_ (false)
+    {}
+
+    virtual void computeParams_impl(const GeometricPath& path,
+        const Vector& gridpoints,
+        Vectors& a, Vectors& b, Vectors& c,
+        Matrices& F, Vectors& g) = 0;
+
+    /// Convert from Collocation to Interpolation
+    /// \todo in this case, the dimensions of the outputs is changed.
+    void collocationToInterpolate (const Vector& gridpoints,
+        Vectors& a, Vectors& b, Vectors& c,
+        Matrices& F, Vectors& g);
+
+    Eigen::Index k_, m_;
+    DiscretizationType discretizationType_;
+    bool constantF_;
+}; // class LinearConstraint
+
+/** BoxConstraint are LinearConstraint of the following form:
+ *  \f{eqnarray}
+ *      x^b_{i, 0} \leq x & \leq x^b_{i, 1} \\
+ *      u^b_{i, 0} \leq u & \leq u^b_{i, 1}
+ *  \f}
+ *  where \f$x^b, u^b\f$ are the Bound vectors.
+ *
+ *  This class handles bounds on \f$x\f$, \f$u\f$ or both.
+ *
+ *  The LinearConstraint representation is computed as follows:
+ *  \f{eqnarray}
+ *      \mathbf a_i &= (1, -1, 0, 0).T \\
+ *      \mathbf b_i &= (0, 0, 1, -1).T \\
+ *      \mathbf c_i &= (-u^b_{i,1}, u^b_{i,0}, -x^b_{i,1}, x^b_{i,0}).T \\
+ *      \mathbf F   &= I_4 \\
+ *      \mathbf g_i &= 0_4 \\
+ *  \f}
+ *
+ * */
+class BoxConstraint : public LinearConstraint {
```

Comment:
I understand why you want to have BoxConstraint, it's true that xbound and ubound can both be formulated in the general form as a Linear constraint.

I don't agree with this design because ubound and xbound were there for the sake of efficiency. Many solvers (qpOASES included) have direct bound on its variables for efficiency reason. I want to take advantage of that. 

1. For the sake of generality, why even use BoxConstraint at all? LinearConstraint is sufficient.

Reply:
> Many solvers (qpOASES included) have direct bound on its variables for efficiency reason.

This is exactly how I understood this and it is why the implementation of `BoxConstraints::computeParams_impl` throws an exception. What I have in mind is that wrappers of solvers which handle box constraints as a special case should do a special case about them. What I mean is, in the wrapper class, there is something like:
```c++
if (dynamic_cast<BoxConstraint*>(constraint)) {
 // treat it as a box constraint
} else {
 // treat it as a general linear constraint
}
```

If it is a problem, we can replace the dynamic_cast by a virtual function. For what it is worth, dynamic_cast makes the code looks dirtier (in my opinion) but the overhead is better than that of a virtual function (I don't think we will be optimizing until this level anyway).

Reply:
I agree that we don't have to optimize to that level.

What really bugs me is the fact that we have two different data structures to represent the LinearConstraint class, while having just one is sufficient. Since this is the interface between the problem definition and the algorithm, I would prefer something simple and straightforward.

That said, suppose there is a kind of constraint that actually evaluates to a combination of **both** the (a,b,c, F, g) form and (xb, ub) form, we will need to produce two constraints objects, one LinearConstraint and one BoxConstraint is that right? Yet, I don't think this is a big problem. Mostly concerned with the complexity of having two interfaces to "choose".

Reply:
> I would prefer something simple and straightforward.

Looking at the Python code, there are plenty of `if a is not None` (same for ubound and xbound). This isn't different from what I propose.

The reasons why I changed it are:
- I realized none of the implemented constraints actually use both (do you have any in mind ?)
- most importantly, in Python, it is very easy to return several values whose type can change dynamically (e.g. F and g). In C++, this is not very easy to do and unexpected. I do not saying the API I implemented is the best. Now, if you would like to change it, we have to agree on the API. Before we do that, please have a look at the first draft implementation of the qpOASES wrapper in implemented just now, on my branch *solver*, in file *src/solver/qpOASES-wrapper.cpp*.

Reply:
> > I would prefer something simple and straightforward.
> 
> Looking at the Python code, there are plenty of `if a is not None` (same for ubound and xbound). This isn't different from what I propose.

There are a lot of if/else precisely because there is a single interface. I had in mind something similar for the C++ code actually (checking for empty matrices). 



> * most importantly, in Python, it is very easy to return several values whose type can change dynamically (e.g. F and g). In C++, this is not very easy to do and unexpected. I do not saying the API I implemented is the best. Now, if you would like to change it, we have to agree on the API. Before we do that, please have a look at the first draft implementation of the qpOASES wrapper in implemented just now, on my branch _solver_, in file _src/solver/qpOASES-wrapper.cpp_.

Maybe I am missing something, but I am not sure what is the problem with having a single interface with both (a,b,c,F,g) form and (ub,xb). Perhaps you can kindly point that out? 

For example, I can just make `ub, xb` have shape (N,0) instead of None and concatenate matrices normally.

Reply:
This forces the allocation of several useless std::vector of N empty matrices and N might be a large number.

I will change the code and have two options to remove the BoxConstraint class:
1. add LinearConstraint::computeBounds,
2. change the API of LinearConstraint::computeParams. If I do this, I think it is better if the reference arguments are allocated by LinearConstraint::computeParams. As a bonus, this would avoid the issue I mentioned here https://github.com/hungpham2511/toppra/pull/83#discussion_r408898612.

Reply:
> This forces the allocation of several useless std::vector of N empty matrices and N might be a large number.
> 
> I will change the code and have two options to remove the BoxConstraint class:
> 
> 1. add LinearConstraint::computeBounds,

Right, so basically merge the methods in BoxConstraint back to LinearConstraint, is that right? I think we will also need a method to check if the (a,b,c,F,g) form exists.

> 2. change the API of LinearConstraint::computeParams. If I do this, I think it is better if the reference arguments are allocated by LinearConstraint::computeParams. As a bonus, this would avoid the issue I mentioned here [#83 (comment)](https://github.com/hungpham2511/toppra/pull/83#discussion_r408898612).

That's right. In fact, we can have a method that computes the required number of rows and columns, which can be used by the solver wrapper to allocate matrices if that is necessary. 

Btw, I think people usually use N ~ 100 to 1000. I know one guy even use N ~ 20.

Reply:
I update the code accordingly.